src/Symfony/Component/HttpClient/NativeHttpClient.php

```diff
@@ -0,0 +1,400 @@
+
+            // Do like curl and browsers: turn POST to GET on 301, 302 and 303
+            if (\in_array($statusCode, [301, 302, 303], true)) {
+                $options = stream_context_get_options($context)['http'];
+
+                if ('POST' === $options['method'] || 303 === $statusCode) {
+                    $options['method'] = 'HEAD' === $options['method'] ? 'HEAD' : 'GET';
+                    $options['content'] = '';
+                    $options['header'] = array_filter($options['header'], static function ($h) {
+                        return 0 !== stripos($h, 'Content-Length:') && 0 !== stripos($h, 'Content-Type:');
+                    });
+
+                    stream_context_set_option($context, ['http' => $options]);
+                }
+            }
+
+            // Resolve the host's IP using the local DNS cache if possible
+            $port = parse_url($url['authority'], PHP_URL_PORT);
+            $info['primary_port'] = $port ? $port : ('http:' === $url['scheme'] ? 80 : 443);
+            $port = $port ? ':'.$port : '';
+            $host = parse_url($url['authority'], PHP_URL_HOST);
+            stream_context_set_option($context, 'ssl', 'peer_name', $host);
+
+            if (null === $ip = $multi->dnsCache[$host] ?? null) {
+                if (!$ip = gethostbynamel($host)) {
+                    throw new TransportException(sprintf('Could not resolve host "%s".', $host));
+                }
+                $info['namelookup_time'] += microtime(true) - $now;
+                $multi->dnsCache[$host] = $ip = $ip[0];
+            }
+
+            $info['primary_ip'] = $ip;
+            $url['authority'] = substr_replace($url['authority'], $ip, -\strlen($host) - \strlen($port), \strlen($host));
+
+            if ($onProgress) {
+                // Notify DNS resolution
+                $onProgress();
+            }
+
+            if (false !== (parse_url($location, PHP_URL_HOST) ?? false)) {
+                // Authorization headers MUST NOT follow except for the initial host name as given in the original URL
+                $rawHeaders = $redirectHeaders['host'] === $host ? $redirectHeaders['with_auth'] : $redirectHeaders['no_auth'];
+                $rawHeaders[] = 'host: '.$host.$port;
+                self::configureHeadersAndProxy($context, $host, $rawHeaders, $proxy, $noProxy);
+            }
+
+            return implode('', $url);
+        };
+
+        $context = stream_context_create($context, ['notification' => $notification]);
+        self::configureHeadersAndProxy($context, $host, $options['raw_headers'], $proxy, $noProxy);
+
+        return new NativeResponse($this->multi, $context, implode('', $url), $options, $gzipEnabled, $info, $resolveRedirect, $onProgress);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function stream($responses, float $timeout = null): ResponseStreamInterface
+    {
+        if ($responses instanceof NativeResponse) {
```

Comment:
Can't these checks be factorized in an abstract class since they are the same than in the `CurlHttpClient` ?

Reply:
4 similar lines don't make a factorization unit.